workspaces/src/worker/impls.rs

```diff
@@ -116,7 +117,25 @@ where
 
     /// View the latest block from the network
     pub async fn view_latest_block(&self) -> anyhow::Result<Block> {
-        self.client().view_block(None).await.map(Into::into)
+        self.view_block(BlockReference::latest()).await
+    }
+
+    /// View the block at the supplied height
+    pub async fn view_block_at_height(&self, height: BlockHeight) -> anyhow::Result<Block> {
+        self.view_block(BlockId::Height(height).into()).await
+    }
+
+    /// View the block at the supplied hash
+    pub async fn view_block_at_hash(&self, hash: CryptoHash) -> anyhow::Result<Block> {
+        self.view_block(BlockId::Hash(hash).into()).await
+    }
+
+    /// View the block from the network given the block reference to it.
+    pub async fn view_block(&self, block_ref: BlockReference) -> anyhow::Result<Block> {
```

Comment:
There doesn't seem to be much use in exposing the other APIs if this is still exported. Possibly an alternative would be to have `view_block_by_id` which takes in that BlockId and then `view_latest_block` which takes in the finality? 

Tough line to draw between function bloat and unergonomic functions in this case I think. Perhaps we could consider doing a builder pattern based on the finality/block id that can be re-used for all functions that reference specific blocks?

Reply:
I should mention there is something I didn't include in `BlockReference` which is `SyncCheckpoint` since I didn't know whether we should expose that one quite yet or not, but not everything will be covered if we just have either finality or block_id related methods, so opting for the most common methods and having a more general one seems best in that case.

> Perhaps we could consider doing a builder pattern based on the finality/block id that can be re-used for all functions that reference specific blocks?

I'm not sure if everything would fit into the builder pattern mold going forward, but let's see for this case, we'd have:
```rust
worker.view_block().await
```
which can just be the default method, where we grab the latest block. Then we can have:
```
worker.view_block()
    .block_hash(...)
    // or
    .finality(Finality::Final)
    .await
```
I like since it does solve the function bloat problem, but would it make sense for us to have a builder just to add a single parameter?

Reply:
Alternative maybe being:
```rust
worker
	// Optional
	.with_block_height()
	// Actual method
	.view_block()
```

So that things are consistent across methods, but I do accept that this might cause too much code bloat. I don't have a strong inclination of what will be best, up to you, just sharing suggestions so we don't build up technical debt

Reply:
yeah this one has so many ways to go about it 😅 

I think potentially, parameters can be added further into these queries so the builder pattern might just be the best use here, in terms of scalability of parameters. I see this already with `view_chunk` which has an extra `shard_id` parameter (which I need to add in a separate PR):
```
worker.view_chunk()
   .block_hash(...)
   .shard_id(0)
```

also, alternative would run into the similar issues of function bloat, especially on the surface level of worker. I like that the action has mainly been first with builders, but might not be a huge requirement there. How about anything that requires a `BlockReference`, we can just do a general builder like so which eliminates some code bloat:
```rust
enum ViewerMethod {
   Block,
   State(PrefixId),
   Code(AccountId),
   ...
}

struct Viewer {
   block_ref: BlockReference,
   method: ViewerMethod,
}
```

Reply:
Or one final option, just expose methods directly on `BlockReference` to make that parameter just easier to use to avoid builder patterns

```rust
impl BlockReference {
	fn code_hash(hash: CodeHash) -> Self {..}
	fn height(..) -> Self {..}
	fn latest() ..
	// idk about this one, maybe separate functions for doomslug and final would be better to avoid exposing finality
	fn latest_with_finality(..)
}
```

And then maybe even obfuscate BlockReference enum type to streamline the API since this would only be used for input (correct?)

Reply:
yeah it's only used with input. Let's go with final option then. Wouldn't add too much code either

Reply:
I guess with this we might want to come up with consistent semantics if this is reused. Is it something `<method>` with the latest default and `<method>_with_reference` for customizing? This is definitely opinionated

Reply:
ahh that part is probably gonna put us back into builders again if we don't want function bloat 😅, considering `<method>_with_unknown_future_parameter` could happen. It's also not API breaking if we were to switch stuff like `view_code` to include the builder pattern:
```
worker.view_code(id)
   .block_ref(...)
   .await
```

So maybe a combination of both would be best? Or maybe we just accept function bloat and call it a day haha